library/src/main/java/com/heinrichreimersoftware/materialintro/app/IntroActivity.java

```diff
@@ -360,23 +360,32 @@ private long calculateScrollDuration(int distance) {
         return Math.round(pageScrollDuration * (distance + Math.sqrt(distance)) / 2);
     }
     
-    public void goToFirstSlide() {
+   public void goToFirstSlide() {
         int currentItem = pager.getCurrentItem();
-        if (currentItem > adapter.getCount() - 1) finishIfNeeded();
+        boolean canGo = false;
+        if (currentItem <= 0) return;
 
-        if (canGoForward(currentItem, true)) {
-            smoothScrollPagerTo(0);
+        while (currentItem > 0 && canGoBackward(currentItem, true)) {
+            currentItem--;
+            canGo = true;
+        }
+        if (canGo) {
+            smoothScrollPagerTo(currentItem);
         } else {
             AnimUtils.applyShakeAnimation(this, buttonNext);
         }
     }
 
     public void goToLastSlide() {
         int currentItem = pager.getCurrentItem();
-        if (currentItem > adapter.getCount() - 1) finishIfNeeded();
-
-        if (canGoForward(currentItem, true)) {
-            smoothScrollPagerTo(adapter.getCount() - 1);
+        int count = getCount() - 1;
+        boolean canGo = false;
+        while (currentItem < count && canGoForward(currentItem, true)) {
+            currentItem++;
+            canGo = true;
+        }
+        if (canGo) {
```

Comment:
Check [L372 comment](https://github.com/HeinrichReimer/material-intro/pull/155/files/34e80b7d012964e46f79553cdb1e9822b17a6545..57cb5d9c332444c442b85abc7087b4213a347c66#r96389088).
Should be implemented equally in this case.